### PR TITLE
Enforce data_frozen for partial searchable snapshot _tier_preference

### PR DESCRIPTION
**Pull request by dakrone**
We already set `data_frozen` for partial searchable snapshots when they are mounted (#70786), and
automatically convert values other than the frozen role automatically for these snapshots when the
metadata is loaded (#71014). This commit makes the `_tier_preference` setting validate that the
setting is *only* `data_frozen` when set on a partial searchable snapshot index.

### CODE REVIEW

**Comment by dakrone** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java [hunk 1]
```diff
@@ -91,10 +91,17 @@ public void validate(String value) {
 
         @Override
         public void validate(String value, Map<Setting<?>, Object> settings) {
-            if (Strings.hasText(value) && SearchableSnapshotsConstants.isPartialSearchableSnapshotIndex(settings) == false) {
-                String[] split = value.split(",");
-                if (Arrays.stream(split).anyMatch(DATA_FROZEN::equals)) {
-                    throw new IllegalArgumentException("[" + DATA_FROZEN + "] tier can only be used for partial searchable snapshots");
+            if (Strings.hasText(value)) {
```
Comment (henningandersen) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:94]: I think we need to either make `_tier_preference` validate that partial searchable snapshots always have the value `data_frozen` (i.e., must be set) or else make that the default when unset.

Ideally, we should remove support for include/exclude/require now that we have the chance too? Can be a separate PR if that is easier.

**Review by henningandersen — COMMENTED** [review]
I would like to strengthen it to require `_tier_preference=data_frozen` for frozen indices (aka partial searchable snapshots).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:94]: > Ideally, we should remove support for include/exclude/require now that we have the chance too? Can be a separate PR if that is easier.

Are you talking about include/exclude/require for `_tier`, or *all* include/exclude/require parameters? If you mean only for tier, I think I'd prefer to work out how we want to treat them for regular indices first (since we talked about removing `_tier` in favor of `_tier_preference` for *all* indices) before special casing them for only frozen indices

Comment (henningandersen) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:94]: I was only referring to include/exclude/require for `_tier`. We can tackle it outside this PR. But it feels odd that the only legal exclude value is `data_frozen` for frozen indices, since that is the last thing you want. When you modify `_tier_preference` validation to require `data_frozen`, the validation have to be different for include/exclude/require vs `_tier_preference` anyway, which is why I was suggesting to tackle this now. But perfectly fine to tackle later too.

Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:103]: > I think we need to either make _tier_preference validate that partial searchable snapshots always have the value data_frozen (i.e., must be set) or else make that the default when unset.

I pushed 9bab290fcd7a1b42fba366c331bd44344899d37f for this

Comment (henningandersen) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:103]: I wonder if there is still a loophole of setting `_tier_preference` to either `""` (empty) or `" "` (white-space), since then the default value will not apply and we end up with a frozen shard with no `_tier_preference`? I think we need to just check for null here to fix that, at least for frozen.

Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java]: Yep, I changed this to be a null comparison

**Review by henningandersen — APPROVED** [review]
LGTM.
